feat(condo): DOMA-9723 new organization creation

nomerdvadcatpyat opened the pull request:
- Added the `/auth/organization` page, which the user sees after registering or logging in if they don't have an existing organization. This page displays an invitation to an organization (if there is one), a sent request to an organization (if there is one), or a form to create a new organization.
- Added a new organization creation form with a tin duplicate check. If an organization with same tin already exists, the user will be prompted to send a request to join it.
- Added `OrganizationEmployeeRequest` notification banner, like organization employee invite
- Added `userPhone` and `userName` in `OrganizationEmployeeRequest` schema to display it in OrganizationEmployeeRequest notification banner
- Updated alert styles

<details>
<summary>Screens</summary>
<img width="1420" alt="изображение" src="https://github.com/user-attachments/assets/3e148896-3f51-4aaa-8cdc-1ed18e842b2c" />
<img width="1432" alt="изображение" src="https://github.com/user-attachments/assets/bf63f554-1316-4988-bf7b-9bc3711a041a" />
<img width="1408" alt="Screenshot 2025-03-26 at 12 49 15" src="https://github.com/user-attachments/assets/b9b1e4e9-3fc0-460a-86b4-d0e8b04c642d" />
<img width="1440" alt="изображение" src="https://github.com/user-attachments/assets/0657b7c8-d10a-4c76-81ed-9d20a372f5dc" />
<img width="618" alt="изображение" src="https://github.com/user-attachments/assets/1bbadbf3-cc41-4413-a848-b4cf1f03fac9" />
<img width="610" alt="изображение" src="https://github.com/user-attachments/assets/98bd3e14-8694-4d96-9d29-2db445905c2d" />
<img width="624" alt="изображение" src="https://github.com/user-attachments/assets/28b851cc-a5f4-43b5-be21-4ff549245250" />

</details>

review: https://review-new-organization-creation-flow-condo.r.doma.ai

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
	- Introduced a new organization creation interface with a dedicated page and form for streamlined employee invitations.
	- Launched interactive notifications that allow users to dismiss them directly via action buttons.
	- Added functionality for accepting or rejecting organization employee requests through a modal interface.
	- Implemented new GraphQL mutations for accepting or rejecting organization employee requests and invites.
	- Enhanced the ability to find organizations by TIN with a new query.
	- Added new fields for user information in organization employee requests, improving request handling.

- **Improvements**
	- Enhanced employee role selection with simplified options and clearer user feedback.
	- Upgraded error messaging and translations for smoother guidance during organization and employee request workflows.
	- Expanded notification management with new properties for better user experience.
	- Improved GraphQL schema to include user-related information for organization employee requests.
	- Enhanced caching strategies for organization employee requests to optimize performance.

- **Chores**
	- Updated the underlying database schema to support additional user information.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

nomerdvadcatpyat commented on apps/condo/lang/en/en.json:
```diff
@@ -2923,6 +2951,43 @@
   "ticket.callRecord.incomingCall": "Incoming call from {phone}",
   "ticket.callRecord.outgoingCall": "Outgoing call for {phone}",
   "ticket.callRecord.speed": "Speed",
+  "ticket.category.classifier.ACCESS.name": "Access",
```
just ran `yarn workspace @app/condo lint-translations --fix`

tolmachev21 commented:
We will have the current form for creating an organization if the user is already an employee of the organization and wants to create a new organization, right?
Then what will the flow of checking the INN and the form for sending a request to join and accepting an invitation to join the organization look like?

nomerdvadcatpyat commented:
> We will have the current form for creating an organization if the user is already an employee of the organization and wants to create a new organization, right? Then what will the flow of checking the INN and the form for sending a request to join and accepting an invitation to join the organization look like?

If the user is already an employee of an organization, they will see the same organization creation modal as before, but with updated styles and a TIN search. If an organization is found by TIN, the user will be prompted to send a request to join it.
I've updated the screens section and added screenshots of this flow.
You can also check it out on the review stand — I included the link in the pull request description.

tolmachev21 commented:
> If an organization is found by TIN, the user will be prompted to send a request to join it. I've updated the screens section and added screenshots of this flow

And if an organization with such a TIN has already been added, then there should be two modals on top of each other?

nomerdvadcatpyat commented:
> > If an organization is found by TIN, the user will be prompted to send a request to join it. I've updated the screens section and added screenshots of this flow
> 
> And if an organization with such a TIN has already been added, then there should be two modals on top of each other?

Yes

tolmachev21 commented on apps/condo/domains/organization/components/CreateOrganizationPageContent.tsx:
```diff
@@ -0,0 +1,362 @@
+            && [OrganizationTypeType.ManagingCompany, OrganizationTypeType.ServiceProvider].includes(invite?.organization?.type)
+
+        if (isAcceptedInvite) {
+            // Evict cache for get actual employee in prefetchOrganizationEmployee
```
That is, the value from the prefetch is placed in the local storage?

nomerdvadcatpyat replied:
Yes, it's a prefetch organization employee request, which will be saved to the cache. Query `getActiveOrganizationEmployee`

tolmachev21 commented on apps/condo/domains/organization/queries/OrganizationEmployeeRequest.graphql:
```diff
@@ -0,0 +1,52 @@
+query getLastUserOrganizationEmployeeRequest ($userId: ID!) {
+    requests: allOrganizationEmployeeRequests (
```
Need add ttl for this request in cacheConfig

nomerdvadcatpyat replied:
Decreased `OrganizationEmployeeRequest` cache to 5 min. Also added `network-only` to queries in `/auth/organization` page to display invites and requests data to new user immediately

tolmachev21 commented on apps/condo/domains/organization/queries/OrganizationEmployeeRole.graphql:
```diff
@@ -0,0 +1,12 @@
+query getOrganizationEmployeeRolesByOrganization ($organizationId: ID!) {
+    roles: allOrganizationEmployeeRoles (
```
^^

nomerdvadcatpyat replied:
Why we need change ttl for roles?

tolmachev21 reviewed (approved):
If no changes are needed in the CC, then that's very cool!

toplenboren commented:
Looks good overall, but I do not really like the margins and offsets 

They loook a bit too big to me, to the point where a scrollbar has to appear on 150% screen scale

![image](https://github.com/user-attachments/assets/124e1ce8-39b9-4efc-8964-fc0225d57f1b)

They also look just a bit bigger then on figma board, it would be super cool if we were to fix that

toplenboren commented on apps/condo/domains/organization/queries/FindOrganizationsByTin.graphql:
I believe its a custom query

nomerdvadcatpyat replied:
Yes, because it's custom query

toplenboren commented on apps/condo/domains/organization/schema/OrganizationEmployeeRequest.js:
```diff
@@ -247,6 +269,18 @@ const OrganizationEmployeeRequest = new GQLListSchema('OrganizationEmployeeReque
                 }
             }
 
+            const userId = get(resolvedData, 'user')
```
We can also double check that authorized user is the one that is in resolvedData, so that I cannot possbily know other people's phone numbers

nomerdvadcatpyat replied:
I think we can pass it as is, because user cannot create/update `OrganizationEmployeeRequest` directly (only soft delete), he can do it only with `SendOrganizationEmployeeRequestService`, where an `authedItem` is passed in the `user` field.
And in many tests in `OrganizationEmployeeRequest` admin client creates requests for other users..

toplenboren replied:
agreed, but I'd have added a second check... If im not mistaken it is practicly free (context.auth.user)

toplenboren commented on apps/condo/domains/organization/schema/OrganizationEmployeeRequest.js:
```diff
@@ -247,6 +269,18 @@ const OrganizationEmployeeRequest = new GQLListSchema('OrganizationEmployeeReque
 
+            const userId = get(resolvedData, 'user')
+            if (userId) {
+                const user = await getByCondition('User', {
```
can we use getById here?

nomerdvadcatpyat replied:
Then you will need to manually process the case with `deletedAt: true`, I will leave it like this

toplenboren commented on apps/condo/domains/organization/schema/OrganizationEmployeeRequest.js:
```diff
@@ -247,6 +269,18 @@ const OrganizationEmployeeRequest = new GQLListSchema('OrganizationEmployeeReque
+                    deletedAt: null,
+                })
+                if (user) {
+                    resolvedData['userName'] = user.name
```
Just a food for thought, might be cool to double check names, maybe not here, but in User model

So that I cannot add links or something spammy

nomerdvadcatpyat replied:
That's a good point. I think it is relevant for all fields where the user can enter text

nomerdvadcatpyat commented on apps/condo/domains/common/components/containers/BaseLayout/Header.tsx:
No, there are no `OrganizationEmployeeRequests` and organization creation screens in cc

nomerdvadcatpyat commented on apps/condo/domains/organization/hooks/useOrganizationInvites.tsx:
```diff
@@ -75,23 +75,26 @@ export const useOrganizationInvites = (organizationTypes: Array<OrganizationType
                         action: () => handleAcceptOrReject(invite, 'reject'),
                         title: RejectMessage,
                         secondary: true,
+                        removeNotificationOnClick: true,
```
Because previously, when you click on the button in notification (action), the notification was closed automatically, but now (in the notification of `OrganizationEmployeeRequest`) there is a logic where it does not need to be closed

nomerdvadcatpyat commented on apps/condo/domains/organization/components/OrganizationEmployeeRequests.tsx:
```diff
@@ -0,0 +1,250 @@
+    const {
+        data: userOrganizationsRequestsData,
+        refetch: refetchOrganizationsRequests,
+    } = useGetRequestsForUserOrganizationsQuery({
```
That's a good point. I will make the request happen only if the role has `canManageOrganizationEmployeeRequests`.

nomerdvadcatpyat replied:
I thought twice and realized that such a check is not necessary. In our access rights for `OrganizationEmployeeRequests` there is already a check that only employees with the `canManageOrganizationEmployeeRequests` right can read

nomerdvadcatpyat commented on apps/condo/domains/organization/schema/FindOrganizationsByTinService.js:
```diff
@@ -34,7 +35,6 @@ const MAX_TOTAL_REQUESTS = 50
 
 // NOTE: we use these TIN codes for testing by users,
```
Why? It's steel relevant

nomerdvadcatpyat commented on apps/condo/pages/_app.tsx:
Here (in ssr), a redirect occurs to the `/auth/organization` page if the user does not have organizations yet and if `Page.skipUserPrefetch` = true  is not specified  (I did not make such a flag separately for the organization like `skipOrganizationPrefetch`. It seems that this should be used in the same places as skip user prefetch). Also here saved a `next` argument, so that after creating an organization, the user can return to the page he wanted to go to

nomerdvadcatpyat replied:
Here a redirect occurs to the `/auth/organization` page if the user does not have organizations yet and if `Page.skipUserPrefetch` = true  is not specified  (I did not make such a flag separately for the organization like `skipOrganizationPrefetch`. It seems that this should be used in the same places as skip user prefetch). Also here saved a `next` argument, so that after creating an organization, the user can return to the page he wanted to go to

nomerdvadcatpyat commented on packages/ui/src/components/Alert/style.less:
```diff
@@ -11,6 +11,9 @@
 @alert-warning-icon-color: @condo-global-color-orange-5;
 @alert-error-bg-color: @condo-global-color-red-1;
 @alert-error-icon-color: @condo-global-color-red-5;
+@alert-icon-size: @condo-global-spacing-24;
```
We use less in ui kit

toplenboren replied:
It is true, but less is a legacy project, so better to not use it :-) we will migrate from it at some point

nomerdvadcatpyat commented on apps/condo/domains/organization/components/CreateOrganizationForm.tsx:
```diff
@@ -0,0 +1,511 @@
+    const SearchByTinLimitMessage = intl.formatMessage({ id: 'organization.createOrganizationForm.searchByTinLimit.message' })
+    const ChatInTelegramMessage = intl.formatMessage({ id: 'organization.createOrganizationForm.supportChat' })
+    const SearchByTinLimitDescription = intl.formatMessage({ id: 'organization.createOrganizationForm.limit.description' }, {
+        chatBotLink: (
+            <SecondaryLink target='_blank' href={HelpRequisites?.support_bot ? `https://t.me/${HelpRequisites.support_bot}` : '#'}>
+                {ChatInTelegramMessage}
+            </SecondaryLink>
+        ),
+    })
```
I'll rename it to `SupportChat`. The translation itself does not contain the word telegram.

nomerdvadcatpyat commented on apps/condo/lang/en/en.json:
```diff
@@ -1586,6 +1592,30 @@
+  "organization.createOrganizationForm.cancelModal.title": "Changed your mind about creating an organization?",
+  "organization.createOrganizationForm.field.name.description": "Enter the official name of the organization. It will be automatically copied into various documents.",
+  "organization.createOrganizationForm.foundOrganizationsModal.sendRequest": "Yes, join",
+  "organization.createOrganizationForm.foundOrganizationsModal.title": "Are you an employee and want to join the organization?",
```
I have already asked chatgpt to make these translations, what needs to be paraphrased here? I think they look fine

toplenboren commented:
> > Looks good overall, but I do not really like the margins and offsets
> > They loook a bit too big to me, to the point where a scrollbar has to appear on 150% screen scale
> > ![image](https://private-user-images.githubusercontent.com/33755274/428639590-124e1ce8-39b9-4efc-8964-fc0225d57f1b.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDM0OTM4MTIsIm5iZiI6MTc0MzQ5MzUxMiwicGF0aCI6Ii8zMzc1NTI3NC80Mjg2Mzk1OTAtMTI0ZTFjZTgtMzliOS00ZWZjLTg5NjQtZmMwMjI1ZDU3ZjFiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MDElMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDAxVDA3NDUxMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTQxZjJlZjAyYWJmNTY3ZDE4MTZhZWM3MWQ1MWYxMzhjNzFiYjM5NjYzY2FmMGM2MmNjOGE5YWQyODgzMmUzNzcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.NMPRnRngiubwFfw40ANP7339wY4oT5EqvTZBpEZI-a4)
> > They also look just a bit bigger then on figma board, it would be super cool if we were to fix that
> 
> The problem was the size of the labels in `Form.item`. Fixed it by passing `size='small'` to the `Form`. <img alt="изображение" width="597" src="https://private-user-images.githubusercontent.com/52532264/428941749-f1f463e2-0cc3-4f28-83f3-22225823ed79.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDM1OTYzMDIsIm5iZiI6MTc0MzU5NjAwMiwicGF0aCI6Ii81MjUzMjI2NC80Mjg5NDE3NDktZjFmNDYzZTItMGNjMy00ZjI4LTgzZjMtMjIyMjU4MjNlZDc5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA0MDIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwNDAyVDEyMTMyMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTBkZTE0ZTY3YmQyYjA5YmU5N2I0NWEyNWE4ZDZhYzU0OTUxZGVjNmUzMGQyYWEzZmQ2MzE5ZTYzODUxYTdkNWYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.cQyYvyU9YogH6ZhJwsI_Tmo03ib6Za4EpNG06x2d0Kc">

Look much much better, good job :-)

toplenboren commented:
Looks good to me

Alllex202 commented on apps/condo/domains/organization/components/CreateOrganizationPageContent.tsx:
```diff
@@ -0,0 +1,355 @@
+                data: {
+                    dv: 1,
+                    sender: getClientSideSenderInfo(),
+                    ...pickBy({ isAccepted, isRejected }),
```
What is the use of `pickBy` here?

nomerdvadcatpyat replied:
For filter only exists properties of object

nomerdvadcatpyat commented on apps/condo/domains/organization/components/CreateOrganizationForm.tsx:
```diff
@@ -0,0 +1,513 @@
+                            label={InnMessage}
+                            rules={validators.tin}
+                        >
+                            <Input/>
```
No placeholder in layouts

nomerdvadcatpyat commented on apps/condo/domains/organization/components/CreateOrganizationForm.tsx:
```diff
@@ -0,0 +1,513 @@
+    const handleSendOrganizationRequest = useCallback(async () => {
+        let lastRequest: SendOrganizationEmployeeRequestMutationResult['data']['request']
+
+        for (const organization of foundOrganizations) {
+            const result = await sendOrganizationEmployeeRequest({
+                variables: {
+                    data: {
+                        dv: 1,
+                        sender: getClientSideSenderInfo(),
+                        organization: { id: organization.id },
+                    },
+                },
+            })
+            lastRequest = result?.data?.request
+        }
```
I think this is not needed. I don't think there will be enough requests to impact the server.

nomerdvadcatpyat commented on apps/condo/domains/organization/components/CreateOrganizationForm.tsx:
```diff
@@ -0,0 +1,513 @@
+
+const SKIP_SEARCH_TINS = [...(unavailableTinsForOrganizationsSearch || []), ...DEFAULT_UNAVAILABLE_TINS]
+
+const prepareValidators = ({ requiredValidator, tinValidator, locale }) => ({
+    name: [requiredValidator],
+    tin: [
+        requiredValidator,
+        tinValidator(locale),
+    ],
+})
```
Copied it from old create organization modal :) Will remove it

nomerdvadcatpyat commented on apps/condo/domains/organization/hooks/useOrganizationEmployeeRequests.tsx:
```diff
@@ -0,0 +1,254 @@
+        ?.filter(Boolean)
+    , [actualEmployeesData?.actualEmployees])
+
+    const {
+        data: userOrganizationsRequestsData,
+        refetch: refetchOrganizationsRequests,
+    } = useGetRequestsForUserOrganizationsQuery({
+        variables: {
+            userId: user?.id,
+            userOrganizationIds,
+        },
+        onError,
+        skip: !isAuthenticated || !employee || isEmployeesLoading,
+    })
```
Looks like there are no difference. Ok, i will change it to `!user`. Mb we can remove `isAuthenticated` from `useAuth` since we use `!!user` in our conditions?

![изображение](https://github.com/user-attachments/assets/9d4b4c55-c838-44b3-9251-50a2a7323dd3)

Alllex202 replied:
We can delete it, but I think it's better to do it in other task.

You can mark this field as deprecated in this pr

nomerdvadcatpyat commented on apps/condo/domains/organization/hooks/useOrganizationEmployeeRequests.tsx:
```diff
@@ -0,0 +1,254 @@
+    const handleAcceptRequest = useCallback(async (values) => {
+        const roleId = values['role']
+
+        if (!activeRequest || !roleId) {
+            console.error('Failed to accept or reject request', activeRequest, roleId)
+            closeChooseRoleModal()
+            return
```
I don't think so. This is an unrealistic case where a user has no `activeRequest` in a modal window that only opens when `activeRequest` is present. The same goes for `roleId` — this is form submission handling where the `roleId` field is required. It's a very strange case, and I only added this handling to prevent the requests below from being executed in such a scenario.

nomerdvadcatpyat commented on apps/condo/domains/organization/hooks/useOrganizationEmployeeRequests.tsx:
```diff
@@ -0,0 +1,254 @@
+    const [activeRequest, setActiveRequest] = useState<ActiveOrganizationRequestType>(null)
+
+    const onError = useMutationErrorHandler()
+    const {
+        data: actualEmployeesData,
+        loading: isEmployeesLoading,
+    } = useGetActualOrganizationEmployeesQuery({
+        variables: { userId },
+        skip: !userId || !persistor,
+    })
```
I see only only one `getActualOrganizationEmployees`, cleared the cache before checking

nomerdvadcatpyat commented on apps/condo/pages/_app.tsx:
```diff
@@ -602,6 +602,21 @@ if (!isDisabledSsr || !isSSR()) {
                     userId: user.id,
                 }))
+
+                if (!skipUserPrefetch && !activeEmployee) {
```
I don't see explicit `return` statement upper. There is a `return` statement if `!skipUserPrefetch`, then if `prefetchAuthOrRedirect` returns `redirectToAuth` and the page does not have `skipRedirectToAuth`. Is it definitely safe to remove this condition?

Alllex202 replied:
We are already in the block when the user is found. We only get here if we don't skip the user preload.
So it seems like this condition is redundant.

That is, we need a user to exist (we have already checked this above), but an employee not to be found (we check it here)

Or maybe I misunderstood something in your idea

nomerdvadcatpyat commented on apps/condo/domains/organization/queries/AcceptOrRejectOrganizationEmployeeRequest.graphql:
```diff
@@ -0,0 +1,5 @@
+mutation acceptOrRejectOrganizationEmployeeRequest ($data: AcceptOrRejectOrganizationEmployeeRequestInput!) {
+    request: acceptOrRejectOrganizationEmployeeRequest (data: $data) {
+        id
+    }
+}
```
It's a accept or reject organization employee request mutation. We get the request information for the organization from the `getRequestsForUserOrganizations` query

nomerdvadcatpyat commented on apps/condo/domains/organization/hooks/useOrganizationInvites.tsx:
```diff
@@ -75,23 +75,26 @@ export const useOrganizationInvites = (organizationTypes: Array<OrganizationType
                         action: () => handleAcceptOrReject(invite, 'reject'),
                         title: RejectMessage,
                         secondary: true,
+                        removeNotificationOnClick: true,
```
I will rename it to `keepNotificationOnClick` and save default behavior

nomerdvadcatpyat commented on apps/condo/pages/auth/organization.tsx:
```diff
@@ -0,0 +1,51 @@
+
+const ORGANIZATION_FORM_WRAPPER_STYLES: CSSProperties = { maxWidth: '350px' }
+
+const CreateOrganizationPage: PageComponentType = () => {
```
Mb move it to `/organization/create`, wyt?

nomerdvadcatpyat commented on apps/condo/domains/common/components/TopNotifications.tsx:
```diff
@@ -1,109 +1,85 @@
-import { InfoCircleFilled } from '@ant-design/icons'
-import { css, jsx } from '@emotion/react'
-import { Affix, Alert, AlertProps, Space } from 'antd'
-import React, { useState } from 'react'
+import { jsx } from '@emotion/react'
```
Since I removed CSS-in-JS in this file, I think we can leave 0 lines with JSX :)
I'll remove it

nomerdvadcatpyat commented on apps/condo/domains/organization/components/OrganizationSelect.tsx:
```diff
@@ -2,31 +2,35 @@ import {
     useGetEmployeeInvitesCountQuery,
 } from '@app/condo/gql'
-import { OrganizationTypeType } from '@app/condo/schema'
 import { Dropdown } from 'antd'
```
Simply changing it doesn’t work; global styles need to be adjusted.  
+ these changes should be made callcenter, but I haven’t callcenter here.  
It’s better to do this in a separate PR.
